### PR TITLE
azurerm_api_management - support for  virtual network integration

### LINKED ISSUE #3667: Feature Request: API Management with virtual networks
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
It should be possible to use a API Management in combination with virtual network. This will make it more comfortable to use just internal backends instead of public backends with IP restrictions or client certificates. So this network configuration is not implemented in azurerm_api_management yet, it is also possible to use this feature with "Developer"  tier now (before it was only available in "Premium" tier).

### New or Affected Resource(s)

* azurerm_api_management

### Potential Terraform Configuration

```hcl
resource "azurerm_api_management" "test" {
  name                = "example-apim"
  location            = "${azurerm_resource_group.test.location}"
  resource_group_name = "${azurerm_resource_group.test.name}"
  publisher_name      = "My Company"
  publisher_email     = "company@terraform.io"

  network_type = "${networkType}"

  network_configuration {
    subnet_id  = "${subnetId}"
    vnet_id    = "${vnetId}"
    subnetname = "${subnetName}"
  }  

  sku {
    name     = "Developer"
    capacity = 1
  }
}
```

This is how it looks like from API prospective:
```
[...]
    "virtualNetworkType": "${networkType}"
    "virtualNetworkConfiguration": {
      "subnetResourceId": "${subnetId}",
      "vnetid": "${vnetId}",
      "subnetname": "${subnetName}"
    },
[...]
```

### References

* https://docs.microsoft.com/en-us/azure/api-management/api-management-using-with-vnet

### PR DESCRIPTION
**Pull request by francescopersico**
The PR adds API Management virtual network integration both internal and external.

Fixes #3667

### CODE REVIEW

**Comment by nsmeef** [comment]
Whats the status of this? Would love to get this in soon :)

**Comment by dansali** [comment]
This will be incredibly helpful for me, looking forward to seeing this put in :)

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go [hunk 1]
```diff
@@ -143,6 +143,32 @@ func resourceArmApiManagementService() *schema.Resource {
...
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"subnet_id": {
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
```
Comment (katbyte) [inline azurerm/internal/services/apimanagement/resource_arm_api_management.go]: Could we validate this is a proper id with `azure.ValidateResourceID`

File: azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go [hunk 2]
```diff
@@ -561,3 +581,41 @@ resource "azurerm_api_management" "test" {
 }
 `, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.Locations.Secondary, data.RandomInteger, data.Locations.Ternary, data.RandomInteger)
 }
+
+func testAccAzureRMApiManagement_virtualNetworkInternal(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
```
Comment (katbyte) [inline azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go]: we'll need a provider block here

File: azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go [hunk 3]
```diff
@@ -561,3 +581,41 @@ resource "azurerm_api_management" "test" {
...
+  name     = "acctestRG-%d"
+  location = "%s"
+}
+
+resource "azurerm_virtual_network" "test" {
+  name                = "acctestVNET-%d"
+  location            = "${azurerm_resource_group.test.location}"
```
Comment (katbyte) [inline azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go]: could we switch the test to .12 syntax?

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go [hunk 4]
```diff
@@ -674,6 +713,8 @@ func resourceArmApiManagementServiceRead(d *schema.ResourceData, meta interface{
 		d.Set("management_api_url", props.ManagementAPIURL)
 		d.Set("scm_url", props.ScmURL)
 		d.Set("public_ip_addresses", props.PublicIPAddresses)
+		d.Set("private_ip_addresses", props.PrivateIPAddresses)
```
Comment (katbyte) [inline azurerm/internal/services/apimanagement/resource_arm_api_management.go]: This property doesn't exist?

Comment (katbyte) [reply azurerm/internal/services/apimanagement/resource_arm_api_management.go]: could we add it to the schema

**Comment by tombuildsstuff** [comment]
@francescopersico @francescopersico unfortunately there's an issue in this version of the API which is fixed in the latest release (but not the intermediate 2019-01-01) release - this is available in the new version of the SDK and as such I'm going to mark this as waiting on that for the moment. Once the SDK been updated we should be able to switch over to this new API version, update the code to match the changes in that version and then update this PR

**Comment by francescopersico** [comment]
@katbyte @tombuildsstuff i have tested and yes there is a problem deleting the resource, the subnet can't be deleted because the APIM deletion process is really long and in background.

**Comment by cs-marek-ruszczyk** [comment]
Hi, I'd like to say if you're going to release this in next version with resource deletion problem or more likely you would like to way for improvements on Azure side?

**Comment by katbyte** [comment]
@francescopersico & @cs-marek-ruszczyk,

We have updated the SDK so we should be able to update the API version used now

**Comment by dansali** [comment]
I'm running into a bunch of Azure related bugs because I'm not able to launch the APIM resource with a VNet configuration applied to it during deployment. E.g. The resource refuses to use a Private Zone DNS in my VNet.

I would love to see this PR merged and released. How are we looking here?

**Comment by francescopersico** [comment]
@katbyte as @tombuildsstuff said the fix is not included in the API v2019-01-01. So i think we must wait for the update of the sdk to v40.6.0 which includes v2019-12-01.
Am i wrong?

**Comment by katbyte** [comment]
@francescopersico, sorry you are correct, i'll try and open a PR for that upgrade later this week after the next release goes out

**Comment by katbyte** [comment]
@francescopersico - i just updated the sdk in #6419 ... however when i vended in the `2019-12-01` api it seems to be missing the `PropertyClient` client =/ didn't go much further then that but looks like there have been a bunch of changes & type renames 😞

**Comment by katbyte** [comment]
@francescopersico, i looked into the new SDK some more and it seems property has been renamed to NamedValue - so we'll need to update the existing resource to migrate to the new ID, deprecate it, and create a new one with the new name 🙁

**Comment by katbyte** [comment]
@francescopersico  - i've opened a PR to update the SDK here #6479

**Comment by katbyte** [comment]
@francescopersico - API update has been merged!

**Comment by jonatasbaldin** [comment]
Excited to see this coming! Just did a PowerShell script inside Terraform to enable vNet in APIM today heheh thanks for allllll the hard work :)

**Comment by francescopersico** [comment]
> @francescopersico - API update has been merged!

Going to update the PR. Thanks.

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go (hunk 1 above)
Comment (francescopersico) [reply azurerm/internal/services/apimanagement/resource_arm_api_management.go:145]: Done!

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go (hunk 4 above)
Comment (francescopersico) [reply azurerm/internal/services/apimanagement/resource_arm_api_management.go:621]: Done!

File: azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go (hunk 2 above)
Comment (francescopersico) [reply azurerm/internal/services/apimanagement/tests/resource_arm_api_management_test.go:536]: Done!

**Comment by francescopersico** [comment]
@katbyte rebase and change requests done.

**Comment by katbyte** [comment]
@francescopersico - i had to add a default to get the tests to pass, but pushed the changes and will merge as soon as travis finishes 🙂

**Comment by dansali** [comment]
Thank you guys!!!

**Comment by dansali** [comment]
@katbyte 

This change also includes a new resource variable "private_ip_addresses" which was missed in the documentation. I think it should also be added in? :)

website/docs/r/api_management.html.markdown

* `private_ip_addresses` - Private Static Load Balanced IP addresses of the API Management service in the additional location. Available only for Basic, Standard and Premium SKU.

**Pull request merged**